linkerd/app/core/src/errors/respond.rs

```diff
@@ -76,76 +121,138 @@ impl<ReqB, RspB: Default + hyper::body::HttpBody>
+        let error = match res {
+            Ok(rsp) => {
+                return Ok(rsp.map(|b| match self {
+                    Respond {
+                        is_grpc: true,
+                        rescue,
+                        ..
+                    } => ResponseBody::GrpcRescue {
+                        inner: b,
+                        trailers: None,
+                        rescue: rescue.clone(),
+                    },
+                    _ => ResponseBody::Passthru(b),
+                }));
+            }
+            Err(error) => error,
+        };
 
-                // Set the l5d error header on all responses.
-                let mut builder = http::Response::builder();
-                builder = set_l5d_proxy_error_header(builder, &*error);
-
-                if self.is_grpc {
-                    let mut rsp = builder
-                        .version(http::Version::HTTP_2)
-                        .header(http::header::CONTENT_LENGTH, "0")
-                        .header(http::header::CONTENT_TYPE, GRPC_CONTENT_TYPE)
-                        .body(ResponseBody::default())
-                        .expect("app::errors response is valid");
-                    let code = set_grpc_status(&*error, rsp.headers_mut());
-                    debug!(?code, "Handling error with gRPC status");
-                    return Ok(rsp);
-                }
+        let span = info_span!(
+            "rescue",
+            client.addr = %self.client
+                .as_ref()
+                .map(|ClientHandle { addr, .. }| *addr)
+                .unwrap_or_else(|| {
+                    tracing::debug!("Missing client address");
+                    ([0, 0, 0, 0], 0).into()
+                })
+        );
 
-                let rsp = set_http_status(builder, &*error)
-                    .version(self.version)
-                    .header(http::header::CONTENT_LENGTH, "0")
-                    .body(ResponseBody::default())
-                    .expect("error response must be valid");
-                let status = rsp.status();
-                debug!(%status, version = ?self.version, "Handling error with HTTP response");
-                Ok(rsp)
+        let SyntheticHttpResponse {
+            grpc_status,
+            http_status,
+            close_connection,
+            message,
+        } = span.in_scope(|| {
+            tracing::info!(%error, "Request failed");
+            self.rescue.rescue(error)
+        })?;
+
+        if close_connection {
+            if let Some(ClientHandle { close, .. }) = self.client.as_ref() {
+                close.close();
+            } else {
+                tracing::debug!("Missing client handle");
             }
         }
+
+        let rsp = if self.is_grpc {
+            Self::grpc_response(grpc_status, &*message)
+        } else {
+            Self::http_response(self.version, http_status, &*message, close_connection)
+        };
+
+        Ok(rsp)
```

Comment:
nit/TIOLI: could golf this down to 
```suggestion
        if self.is_grpc {
            Ok(Self::grpc_response(grpc_status, &*message))
        } else {
            Ok(Self::http_response(self.version, http_status, &*message, close_connection))
        }
```

or even
```suggestion
        Ok(if self.is_grpc {
            Self::grpc_response(grpc_status, &*message)
        } else {
             Self::http_response(self.version, http_status, &*message, close_connection)
        })
```
although maybe that's less readable?